tests/phpunit/E2E/Core/UserOverrideTest.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace E2E\Core;
+
+/**
+ * Class UserOverrideTest
+ *
+ * Check that overriding session user behaves as expected.
+ *
+ * @package E2E\Core
+ * @group e2e
+ */
+class UserOverrideTest extends \CiviEndToEndTestCase {
+
+  protected function setUp() {
+    parent::setUp();
+  }
+
+  protected function tearDown() {
+    while (\CRM_Core_Session::singleton()->restoreCurrentUser()) {
+      // Loop until all overrides are cleared
+    }
+  }
+
+  public function testOverride() {
+    $session = \CRM_Core_Session::singleton();
+    $originalUser = $session::getLoggedInContactID();
```

Comment:
I'm a little surprised PHP 7 doesn't complain about dispatching a static-function on an object. Learn something new...

Reply:
It's not really different from doing `self::fooParentFunction()` from within the object.